File: python/src/emay_sleepo2/protocol.py
"""Pure-protocol layer for the EMAY SleepO2 BLE protocol.

Contains no Bluetooth calls, no async, no platform dependencies. Each
language binding reimplements these functions with that language's
byte-manipulation idioms.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import List, Optional

from .types import Reading

# ---- BLE identifiers ----
SERVICE_UUID = "0000ff12-0000-1000-8000-00805f9b34fb"
WRITE_UUID = "0000ff01-0000-1000-8000-00805f9b34fb"
NOTIFY_UUID = "0000ff02-0000-1000-8000-00805f9b34fb"
NAME_PREFIX = "SleepO2"

# ---- Commands ----
HELLO: bytes = bytes([0x89, 0x09])
DEVICE_STATE: bytes = bytes([0x8E, 0x05, 0x13])
START_REALTIME: bytes = bytes([0x9B, 0x01, 0x1C])
STOP_REALTIME: bytes = bytes([0x9B, 0x7F, 0x1A])
GET_BATTERY: bytes = bytes([0x86, 0x06])
HEARTBEAT: bytes = bytes([0x9A, 0x1A])

START_SEQUENCE: List[bytes] = [HELLO, DEVICE_STATE, START_REALTIME, GET_BATTERY]

# ---- Data frame constants ----
FRAME_LENGTH: int = 8
FRAME_HEADER: bytes = bytes([0xEB, 0x01, 0x05])
FRAME_TRAILER: bytes = bytes([0x7F, 0x00])

# ---- Plausibility bounds ----
PULSE_MIN_BPM: int = 30
PULSE_MAX_BPM: int = 220
SPO2_MAX_PERCENT: int = 100
SPO2_MIN_PERCENT: int = 0

# ---- Sentinel values ----
_SENTINEL_VALUES = {0x00, 0xFF}


def checksum(payload: bytes) -> int:
    """Compute the EMAY checksum: sum(payload) & 0x7F.

    The 0x7F mask (NOT 0xFF) is crucial — 0xFF produces invalid
    checksums that the device silently ignores.
    """
    return sum(payload) & 0x7F


def command(payload: bytes) -> bytes:
    """Build a full command frame: payload + checksum."""
    return payload + bytes([checksum(payload)])


def parse_reading(raw: bytes) -> Optional[Reading]:
    """Attempt to parse an 8-byte raw frame from the BLE notify characteristic.

    Returns a Reading on success, or None if the frame
    fails any validation check. Both spo2 and pulse can individually be
    None when the sensor reports "no finger detected" for that metric.
    """
    if len(raw) != FRAME_LENGTH:
        return None
    if raw[0] != FRAME_HEADER[0] or raw[1] != FRAME_HEADER[1] or raw[2] != FRAME_HEADER[2]:
        return None
    if raw[5] != FRAME_TRAILER[0] or raw[6] != FRAME_TRAILER[1]:
        return None

    cks = sum(raw[:7]) & 0x7F
    if raw[7] != cks:
        return None

    raw_pr = raw[3]
    raw_spo2 = raw[4]

    pr: Optional[int] = None if raw_pr in _SENTINEL_VALUES else raw_pr
    spo2: Optional[int] = None if raw_spo2 in _SENTINEL_VALUES else raw_spo2

    # Application-level plausibility bounds
    if pr is not None and (pr < PULSE_MIN_BPM or pr > PULSE_MAX_BPM):
        return None
    if spo2 is not None and (spo2 < SPO2_MIN_PERCENT or spo2 > SPO2_MAX_PERCENT):
        return None

    return Reading(spo2=spo2, pulse=pr, timestamp=datetime.now(timezone.utc))

File: rust/src/protocol.rs
/// Pure-protocol layer for the EMAY SleepO2 BLE protocol.
/// no_std compatible with alloc.
use crate::types::Reading;

// ---- BLE identifiers ----
pub const SERVICE_UUID: &str = "0000ff12-0000-1000-8000-00805f9b34fb";
pub const WRITE_UUID: &str = "0000ff01-0000-1000-8000-00805f9b34fb";
pub const NOTIFY_UUID: &str = "0000ff02-0000-1000-8000-00805f9b34fb";
pub const NAME_PREFIX: &str = "SleepO2";

// ---- Command bytes ----
pub const HELLO: [u8; 2] = [0x89, 0x09];
pub const DEVICE_STATE: [u8; 3] = [0x8E, 0x05, 0x13];
pub const START_REALTIME: [u8; 3] = [0x9B, 0x01, 0x1C];
pub const STOP_REALTIME: [u8; 3] = [0x9B, 0x7F, 0x1A];
pub const GET_BATTERY: [u8; 2] = [0x86, 0x06];
pub const HEARTBEAT: [u8; 2] = [0x9A, 0x1A];

pub const START_SEQUENCE: &[&[u8]] = &[&HELLO, &DEVICE_STATE, &START_REALTIME, &GET_BATTERY];

// ---- Data frame constants ----
const FRAME_LENGTH: usize = 8;
const FRAME_HEADER: [u8; 3] = [0xEB, 0x01, 0x05];
const FRAME_TRAILER: [u8; 2] = [0x7F, 0x00];

// ---- Plausibility bounds ----
const PULSE_MIN_BPM: u8 = 30;
const PULSE_MAX_BPM: u8 = 220;
const SPO2_MIN_PCT: u8 = 0;
const SPO2_MAX_PCT: u8 = 100;
const SENTINEL_VALUES: [u8; 2] = [0x00, 0xFF];

/// Compute the EMAY checksum: sum(payload) & 0x7F.
pub fn checksum(payload: &[u8]) -> u8 {
    (payload.iter().map(|&b| b as u16).sum::<u16>() & 0x7F) as u8
}

/// Build a full command frame: payload + checksum.
pub fn command(payload: &[u8]) -> Vec<u8> {
    let mut v = payload.to_vec();
    v.push(checksum(payload));
    v
}

/// Attempt to parse an 8-byte raw frame. Returns Some(Reading) on
/// success, None if any validation fails.
pub fn parse_reading(raw: &[u8]) -> Option<Reading> {
    if raw.len() != FRAME_LENGTH { return None; }
    if raw[0] != FRAME_HEADER[0] || raw[1] != FRAME_HEADER[1] || raw[2] != FRAME_HEADER[2] {
        return None;
    }
    if raw[5] != FRAME_TRAILER[0] || raw[6] != FRAME_TRAILER[1] {
        return None;
    }
    let cks = (raw[..7].iter().map(|&b| b as u16).sum::<u16>() & 0x7F) as u8;
    if raw[7] != cks { return None; }

    let raw_pr = raw[3];
    let raw_spo2 = raw[4];

    let is_sentinel = |b: u8| SENTINEL_VALUES.contains(&b);

    let pulse = if is_sentinel(raw_pr) { None } else { Some(raw_pr) };
    let spo2 = if is_sentinel(raw_spo2) { None } else { Some(raw_spo2) };

    if let Some(p) = pulse { if p < PULSE_MIN_BPM || p > PULSE_MAX_BPM { return None; } }
    if let Some(s) = spo2 { if s < SPO2_MIN_PCT || s > SPO2_MAX_PCT { return None; } }

    Some(Reading::new(spo2, pulse))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn frame(pr: u8, spo2: u8) -> Vec<u8> {
        let mut buf = vec![0xEB, 0x01, 0x05, pr, spo2, 0x7F, 0x00, 0x00];
        let ck: u8 = (buf[..7].iter().map(|&b| b as u16).sum::<u16>() & 0x7F) as u8;
        buf[7] = ck;
        buf
    }

    #[test]
    fn test_checksums() {
        assert_eq!(checksum(&[0x89]), 0x09);
        assert_eq!(checksum(&[0x9A]), 0x1A);
        assert_eq!(checksum(&[0x9B, 0x01]), 0x1C);
        assert_eq!(checksum(&[0x8E, 0x05]), 0x13);
        assert_eq!(checksum(&[0x9B, 0x7F]), 0x1A);
    }

    #[test]
    fn test_valid_frame() {
        let r = parse_reading(&frame(62, 98)).unwrap();
        assert_eq!(r.pulse, Some(62));
        assert_eq!(r.spo2, Some(98));
    }

    #[test]
    fn test_wrong_length() {
        assert!(parse_reading(&[0xEB, 0x01]).is_none());
    }

    #[test]
    fn test_bad_checksum() {
        let bad = vec![0xEB, 0x01, 0x05, 62, 98, 0x7F, 0x00, 0xFF];
        assert!(parse_reading(&bad).is_none());
    }

    #[test]
    fn test_sentinel_pr_0() {
        let r = parse_reading(&frame(0x00, 98)).unwrap();
        assert_eq!(r.pulse, None);
        assert_eq!(r.spo2, Some(98));
    }

    #[test]
    fn test_sentinel_spo2_ff() {
        let r = parse_reading(&frame(62, 0xFF)).unwrap();
        assert_eq!(r.spo2, None);
    }

    #[test]
    fn test_both_sentinels() {
        let r = parse_reading(&frame(0xFF, 0xFF)).unwrap();
        assert_eq!(r.pulse, None);
        assert_eq!(r.spo2, None);
    }

    #[test]
    fn test_pulse_too_low() {
        assert!(parse_reading(&frame(29, 98)).is_none());
    }

    #[test]
    fn test_pulse_too_high() {
        assert!(parse_reading(&frame(221, 98)).is_none());
    }

    #[test]
    fn test_pulse_boundary() {
        assert_eq!(parse_reading(&frame(30, 98)).unwrap().pulse, Some(30));
        assert_eq!(parse_reading(&frame(220, 98)).unwrap().pulse, Some(220));
    }

    #[test]
    fn test_spo2_too_high() {
        assert!(parse_reading(&frame(62, 101)).is_none());
    }

    #[test]
    fn test_spo2_boundary() {
        assert_eq!(parse_reading(&frame(62, 0)).unwrap().spo2, None);
        assert_eq!(parse_reading(&frame(62, 1)).unwrap().spo2, Some(1));
        assert_eq!(parse_reading(&frame(62, 100)).unwrap().spo2, Some(100));
    }
}
